src/rpc/server/connection.zig

```diff
@@ -0,0 +1,264 @@
+const builtin = @import("builtin");
+const std = @import("std");
+
+pub const GetSockNameError = std.posix.GetSockNameError;
+pub fn getSockName(
+    socket_handle: std.posix.socket_t,
+) GetSockNameError!std.net.Address {
+    var addr: std.net.Address = .{ .any = undefined };
+    var addr_len: std.posix.socklen_t = @sizeOf(@TypeOf(addr.any));
+    try std.posix.getsockname(socket_handle, &addr.any, &addr_len);
+    return addr;
+}
+
+/// When this is false, it means `accept[Handled]` can't apply
+/// flags to the accepted socket, and the caller will have to
+/// to ensure relevant flags are enabled/disabled after acceptance.
+pub const have_accept4 = !builtin.target.isDarwin();
+
+pub const AcceptHandledError = HandleAcceptError || error{
+    ConnectionAborted,
+    ProtocolFailure,
+    WouldBlock,
+};
+
+pub fn acceptHandled(
+    tcp_server: std.net.Server,
+    /// NOTE: this is *only* a hint, and may not apply on all platforms.
+    /// See `have_accept4`.
+    sync_hint: enum { blocking, nonblocking },
+) AcceptHandledError!std.net.Server.Connection {
+    var accept_flags: u32 = std.posix.SOCK.CLOEXEC;
+    accept_flags |= switch (sync_hint) {
+        .blocking => 0,
+        .nonblocking => std.posix.SOCK.NONBLOCK,
+    };
+
+    while (true) {
+        var addr: std.net.Address = .{ .any = undefined };
+        var addr_len: std.posix.socklen_t = @sizeOf(@TypeOf(addr.any));
+        const rc = if (have_accept4)
+            std.posix.system.accept4(
+                tcp_server.stream.handle,
+                &addr.any,
+                &addr_len,
+                accept_flags,
+            )
+        else
+            std.posix.system.accept(
+                tcp_server.stream.handle,
+                &addr.any,
+                &addr_len,
+            );
+
+        return switch (try handleAcceptResult(std.posix.errno(rc))) {
+            .intr => continue,
+            .conn_aborted => error.ConnectionAborted,
+            .proto_fail => error.ProtocolFailure,
+            .again => error.WouldBlock,
+            .success => .{
+                .stream = .{ .handle = rc },
+                .address = addr,
+            },
+        };
+    }
+}
+
+pub const SetSocketSync = std.posix.FcntlError;
+
+/// Ensure the socket is set to be blocking or nonblocking.
+pub fn setSocketSync(
+    socket: std.posix.socket_t,
+    sync: enum { blocking, nonblocking },
+) !void {
```

Comment:
is this import meant to be like this? could we also keep all imports at the top of the file

Reply:
It isn't an import so much as it is a re-export; `SetSocketSync` is to be interpreted as the error set of of `setSocketSync` - though should be `SetSocketSyncError`.